tests/test_pagination.py

```diff
@@ -152,16 +150,42 @@ def pagination_test(self):
                                     msg="A paginated synced stream has a record that is missing automatic fields.")
 
                     # Verify we have more fields sent to the target than just automatic fields (this is set above)
-                    self.assertEqual(auto_fields.difference(actual_keys),
-                                     set(), msg="A paginated synced stream has a record that is missing expected fields.")
+                    self.assertEqual(set(), auto_fields.difference(actual_keys),
+                                     msg="A paginated synced stream has a record that is missing expected fields.")
 
-                # TODO ADD CHECK ON IDS
-                # Verify by pks that the data replicated matches what we expect
+                primary_keys = self.expected_primary_keys().get(stream)
+                pks = list(primary_keys) if primary_keys else None
+                if not pks:
+                    pks = self.makeshift_primary_keys().get(stream)
 
+                # Verify by pks that the replicated records match our expectations
+                self.assertRecordsEqualByPK(stream, expected_records.get(stream), actual_records, pks)
+
+                # Verify the expected number of records were replicated
+                self.assertEqual(len(expected_records.get(stream)), len(actual_records))
+
+
+    def assertRecordsEqualByPK(self, stream, expected_records, actual_records, pks):
+        """Compare expected and actual records by their primary key."""
+
+        # Verify there are no duplicate pks in the target
+        actual_pks = [tuple(actual_record.get(pk) for pk in pks) for actual_record in actual_records]
+        actual_pks_set = set(actual_pks)
+        self.assertEqual(len(actual_pks), len(actual_pks_set), msg="A duplicate record may have been replicated.")
+
+        # Verify there are no duplicate pks in our expectations
+        expected_pks = [tuple(expected_record.get(pk) for pk in pks) for expected_record in expected_records]
+        expected_pks_set = set(expected_pks)
+        self.assertEqual(len(expected_pks), len(expected_pks_set), msg="Our expectations contain a duplicate record.")
+
+        # Verify that all expected records were replicated
+        self.assertEqual(set(), expected_pks_set.difference(actual_pks_set),
+                         msg="Our expectations have more unique records than the tap replicated.")
+
+        # Verify ONLY expected records were replicated
+        self.assertEqual(set(), actual_pks_set.difference(expected_pks_set),
+                         msg="The tap replicated data that was not in our expectations.")
```

Comment:
Can you compare both sets are equal instead of comparing the difference is 0 both depending on whether you put one first or the other?

Reply:
I liked having the separate error messages, but I guess it is cleaner to have one assertion.